pgrx/src/callconv.rs

```diff
@@ -283,3 +286,190 @@ where
+        // Safety: User must supply a valid fcinfo to assume_valid() in order
+        // to construct a FcInfo. If that constraint is maintained, this should
+        // be safe.
+        unsafe {
+            let mut flinfo = NonNull::new((*self.0).flinfo).unwrap();
+            if flinfo.as_ref().fn_extra.is_null() {
+                flinfo.as_mut().fn_extra = PgMemoryContexts::For(flinfo.as_ref().fn_mcxt)
+                    .leak_and_drop_on_delete(default())
+                    as crate::void_mut_ptr;
+            }
+
+            // Safety: can new_unchecked() here because we just initialized it.
+            NonNull::new_unchecked(flinfo.as_mut().fn_extra as *mut pg_sys::FuncCallContext)
+        }
+    }
+
+    #[inline]
+    pub fn srf_is_initialized(&self) -> bool {
+        // Safety: User must supply a valid fcinfo to assume_valid() in order
+        // to construct a FcInfo. If that constraint is maintained, this should
+        // be safe.
+        unsafe { !(*(*self.0).flinfo).fn_extra.is_null() }
+    }
+
+    /// Thin wrapper around [`pg_sys::init_MultiFuncCall`], made necessary
+    /// because this structure's FunctionCallInfo is a private field.
+    ///
+    /// This should initialize `self.0.flinfo.fn_extra`
+    #[inline]
+    pub unsafe fn init_multi_func_call(&mut self) -> &'fcx mut pg_sys::FuncCallContext {
+        unsafe {
+            let fcx: *mut pg_sys::FuncCallContext = pg_sys::init_MultiFuncCall(self.0);
+            debug_assert!(fcx.is_null() == false);
+            &mut *fcx
+        }
+    }
+
+    /// Equivalent to "per_MultiFuncCall" with no FFI cost, and a lifetime
+    /// constraint.
+    ///
+    /// Safety: Assumes `self.0.flinfo.fn_extra` is non-null
+    /// i.e. [`FcInfo::srf_is_initialized()`] would be `true`.
+    pub(crate) unsafe fn deref_fcx(&mut self) -> &'fcx mut pg_sys::FuncCallContext {
+        unsafe {
+            let fcx: *mut pg_sys::FuncCallContext = (*(*self.0).flinfo).fn_extra.cast();
+            debug_assert!(fcx.is_null() == false);
+            &mut *fcx
+        }
+    }
+
+    /// Safety: Assumes `self.0.flinfo.fn_extra` is non-null
+    /// i.e. [`FcInfo::srf_is_initialized()`] would be `true`.
+    #[inline]
+    pub unsafe fn srf_return_next(&mut self) {
+        unsafe {
+            self.deref_fcx().call_cntr += 1;
+            (*((*self.0).resultinfo as *mut pg_sys::ReturnSetInfo)).isDone =
+                pg_sys::ExprDoneCond_ExprMultipleResult;
+        }
+    }
+
+    /// Safety: Assumes `self.0.flinfo.fn_extra` is non-null
+    /// i.e. [`FcInfo::srf_is_initialized()`] would be `true`.
+    #[inline]
+    pub unsafe fn srf_return_done(&mut self) {
+        unsafe {
+            pg_sys::end_MultiFuncCall(self.0, self.deref_fcx());
+            (*self.get_result_info()).isDone =
+                pg_sys::ExprDoneCond_ExprEndResult;
+        }
+    }
+
+    pub unsafe fn get_result_info(&self) -> *mut pg_sys::ReturnSetInfo { 
```

Comment:
```suggestion
    /// # Safety
    /// Do not corrupt the `pg_sys::ReturnSetInfo` struct's data.
    pub unsafe fn get_result_info(&mut self) -> &mut pg_sys::ReturnSetInfo { 
```

Alternately, expose safe accessors for just the last four fields (they are "owned" by the callee):

https://github.com/postgres/postgres/blob/4ebe51a5fbc3e08861db85fe3868c31c6bc8ed59/src/include/nodes/execnodes.h#L335-L341

Reply:
or just the one we set.

Reply:
Which one do we mutate?

Reply:
seems so?

Reply:
Alright, implemented in 9b420292aae38b1beaa324f067378cce89ee6b90